### ISSUE TITLE
`incompatible_msrv` does not work with unit tests

### ISSUE THREAD

**Issue by axelkar** [rust-lang/rust-clippy#15369]
### Summary

`incompatible_msrv` does not seem to work with unit tests when it's gated to `cfg(test)`, even with `--tests`.

The [documentation of the lint](https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv) recommend using inner attributes, I'm using outer because I'm on Rust stable, but it shouldn't make any difference.

> ```rs
> #![cfg_attr(test, clippy::msrv = "1.85.0")]
> ```

### Reproducer

See comments:

```rust
#[cfg(test)]
// I was going to set this higher from the library's own MSRV but I wanted to see that it worked
#[clippy::msrv = "1.60.0"]
mod tests {
    use std::sync::LazyLock;

    static _FOO: LazyLock<()> = LazyLock::new(|| ()); // Clippy lints as clippy::incompatible_msrv only without #[cfg(test)]
    static _BAR: () = {
        (); // Clippy lints as clippy::no_effect with and without #[cfg(test)]
    };

    #[test]
    fn bar() {
        LazyLock::new(|| ()); // No lint even without #[cfg(test)]
        (); // Clippy lints as clippy::no_effect with and without #[cfg(test)]
    }
}
```

### Version

```text
rustc 1.87.0 (17067e9ac 2025-05-09) (built from a source tarball)
binary: rustc
commit-hash: 17067e9ac6d7ecb70e50f92c1944e545188d2359
commit-date: 2025-05-09
host: x86_64-unknown-linux-gnu
release: 1.87.0
LLVM version: 20.1.6
```

### Additional Labels

@rustbot label +I-false-negative
@rustbot label +L-tests
@rustbot label +L-suspicious

**Comment by profetia** [comment]
There is a `check-incompatible-msrv-in-tests` to config this.

**Comment by axelkar** [comment]
Thanks for sharing that! Without it there's some unexpected behavior.

|  [^msrv-config] | gated by `cfg(test)` | `--all-targets`? | [^static-msrv] | [^static-no-effect] | [^test-msrv] | [^test-no-effect] | Notes |
| - | - | - | - | - | - | - | - |
| yes | yes | yes | x | x | x | x |
| yes | no  | yes | x | x | x | x |
| no  | yes | yes |  | x |   | x |
| no  | no  | yes | x | x |   | x | <- Unexpected behavior caused by https://github.com/rust-lang/rust-clippy/pull/12261 |


For completeness, without `--all-targets`:
|  [^msrv-config] | gated by `cfg(test)` | `--all-targets`? | [^static-msrv] | [^static-no-effect] | [^test-msrv] | [^test-no-effect] | Notes |
| - | - | - | - | - | - | - | - |
| yes | yes | no  |   |   |   |   |
| yes | no  | no  | x | x |  |   |
| no  | yes | no  |   |   |   |  |
| no  | no  | no  | x | x |   |   |

[^msrv-config]: `check-incompatible-msrv-in-tests`
[^static-msrv]: Lint `incompatible-msrv` emitted in `static`
[^static-no-effect]: Lint `no-effect` emitted in `static`
[^test-msrv]: Lint `incompatible-msrv` emitted in `#[test]`
[^test-no-effect]: Lint `no-effect` emitted in `#[test]`

**Comment by tamird** [comment]
cc @samueltardieu
